derive/src/type_def.rs

```diff
@@ -14,19 +14,160 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use crate::impl_wrapper::wrap;
 use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
 use syn::{
-    self,
-    parse::Result,
+	self, parse::Result, parse_quote, punctuated::Punctuated, token::Comma, Data, DataEnum, DataStruct, DataUnion,
+	DeriveInput, Expr, ExprLit, Field, Fields, Lit, Variant,
 };
 
 pub fn generate(input: TokenStream2) -> TokenStream2 {
-    match generate_impl(input.into()) {
-        Ok(output) => output.into(),
-        Err(err) => err.to_compile_error().into(),
-    }
+	match generate_impl(input.into()) {
+		Ok(output) => output.into(),
+		Err(err) => err.to_compile_error().into(),
+	}
 }
 
-pub fn generate_impl(_input: TokenStream2) -> Result<TokenStream2> {
-	unimplemented!()
+pub fn generate_impl(input: TokenStream2) -> Result<TokenStream2> {
+	let mut ast: DeriveInput = syn::parse2(input)?;
+
+	// add bound
+	ast.generics.type_params_mut().for_each(|p| {
+		p.bounds.push(parse_quote!(_type_metadata::HasTypeId));
+	});
+
+	let ident = &ast.ident;
+	let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+
+	let def = match &ast.data {
+		Data::Struct(ref s) => generate_struct_def(s),
+		Data::Enum(ref e) => generate_enum_def(e),
+		Data::Union(ref u) => generate_union_def(u),
+	};
+
+	let has_type_def_impl = quote! {
+		impl #impl_generics _type_metadata::HasTypeDef for #ident #ty_generics #where_clause {
+			fn type_def() -> _type_metadata::TypeDef {
+				#def.into()
+			}
+		}
+	};
+
+	Ok(wrap(ident, "HAS_TYPE_DEF", has_type_def_impl).into())
+}
+
+type FieldsList = Punctuated<Field, Comma>;
+
+fn generate_fields_def(fields: &FieldsList) -> TokenStream2 {
+	let fields_def = fields.iter().map(|f| {
+		let (ty, ident) = (&f.ty, &f.ident);
+		if let Some(i) = ident {
+			let type_id = quote! {
+				<#ty as _type_metadata::HasTypeId>::type_id()
+			};
+			quote! {
+				_type_metadata::NamedField::new(stringify!(#i), #type_id)
+			}
+		} else {
+			quote! {
+				_type_metadata::UnnamedField::new::<#ty>()
```

Comment:
`#ty` is probably wrong here. The `UnnamedField` expects a `Form` type and in this case you can leave it away because you return a `FreeForm` which is the default.

Reply:
Not sure if I follow, you mean we should make it be able to use other `Form` type in this impl? (Btw currently `Form` types are private).

Reply:
Ah damn I was confusing the `new` generic param with the `UnnamedField` generic `Form` param ... my bad!